signer/signer.go

```diff
@@ -0,0 +1,127 @@
+// Package signer provides notation signing functionality. It implements the
+// notation.Signer interface by providing builtinSigner for local signing and
+// pluginSigner for remote signing.
+package signer
+
+import (
+	"context"
+	"crypto"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/notaryproject/notation-core-go/signature"
+	"github.com/notaryproject/notation-go"
+	"github.com/notaryproject/notation-go/internal/envelope"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+// mediaTypePayloadV1 is the supported content type for signature's payload.
+const mediaTypePayloadV1 = "application/vnd.cncf.notary.payload.v1+json"
+
+// signingAgent is the unprotected header field used by signature.
+const signingAgent = "Notation/1.0.0"
+
+// builtinSigner implements notation.Signer and wraps LocalSigner for signing
+// locally
+type builtinSigner struct {
+	signature.LocalSigner
+}
+
+// New returns a builtinSigner given key and cert chain
+func New(key crypto.PrivateKey, certChain []*x509.Certificate) (notation.Signer, error) {
+	localSigner, err := signature.NewLocalSigner(certChain, key)
+	if err != nil {
+		return nil, err
+	}
+	return &builtinSigner{
+		LocalSigner: localSigner,
+	}, nil
+}
+
+// NewFromFiles returns a builtinSigner given key and certChain paths.
+func NewFromFiles(keyPath, certChainPath string) (notation.Signer, error) {
+	if keyPath == "" {
+		return nil, errors.New("key path not specified")
+	}
+	if certChainPath == "" {
+		return nil, errors.New("certificate path not specified")
+	}
+
+	// read key / cert pair
+	cert, err := tls.LoadX509KeyPair(certChainPath, keyPath)
+	if err != nil {
+		return nil, err
+	}
+	if len(cert.Certificate) == 0 {
+		return nil, fmt.Errorf("%q does not contain a signer certificate chain", certChainPath)
+	}
+
+	// parse cert
+	certs := make([]*x509.Certificate, len(cert.Certificate))
+	for i, c := range cert.Certificate {
+		certs[i], err = x509.ParseCertificate(c)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// create signer
+	return New(cert.PrivateKey, certs)
+}
+
+// Sign signs the artifact described by its descriptor and returns the
+// marshaled envelope.
+func (s *builtinSigner) Sign(ctx context.Context, desc ocispec.Descriptor, opts notation.SignOptions) ([]byte, *signature.SignerInfo, error) {
+	return generateSignatureBlob(ctx, s.LocalSigner, desc, opts)
+}
+
+func generateSignatureBlob(ctx context.Context, signer signature.Signer, desc ocispec.Descriptor, opts notation.SignOptions) ([]byte, *signature.SignerInfo, error) {
```

Comment:
The parameter `ctx` is never used.

Reply:
This is useful when we want to add logger in the future. Should we keep it here?